File: tenet-py/tenet/codegen.py
"""代码生成器（Codegen）：把 Tenet AST 编译成 Go 源码。

与 Rust 版 tenet-rs/src/codegen.rs 完全对应——生成结果**逐字节一致**：

| Tenet | Go |
|-------|----|
| `int` / `float` / `bool` / `string` | `int64` / `float64` / `bool` / `string` |
| `let x: T = v;` | `var x T = v;` |
| `while (c) { }` | `for (c) { }`（Go 没有 while） |
| `print(...)` | `fmt.Println(...)` |
| `fn f(a: int) -> int { }` | `func f(a int64) int64 { }` |

`let` 省略类型标注时由 `infer_type` 静态推断，规则与解释器语义一致。
"""

from __future__ import annotations

from . import ast
from . import value as val
from .error import TenetError

_GO_TYPE = {
    ast.T_INT: "int64",
    ast.T_FLOAT: "float64",
    ast.T_BOOL: "bool",
    ast.T_STR: "string",
}


def go_type(t: str) -> str:
    return _GO_TYPE[t]


def go_binop(op: str) -> str:
    return op  # Tenet 与 Go 的运算符符号完全一致


class GoCodegen:
    def __init__(self):
        self.out: list[str] = []
        self.indent = 0
        # 当前作用域内可见的变量类型
        self.symbols: dict[str, str] = {}
        # 全局函数签名表：name -> (返回类型或 None)
        self.functions: dict[str, str | None] = {}

    @classmethod
    def generate(cls, program: ast.Program) -> str:
        gen = cls()
        gen.collect_signatures(program)
        gen.emit_program(program)
        return "".join(gen.out)

    # ---- 输出辅助（与 Rust push 行为一致）----

    def push(self, line: str) -> None:
        if line == "":
            self.out.append("\n")
            return
        self.out.append("\t" * self.indent + line + "\n")

    # ---- 预处理 ----

    def collect_signatures(self, program: ast.Program) -> None:
        for stmt in program.stmts:
            if isinstance(stmt, ast.FnDecl):
                if stmt.name in self.functions:
                    raise TenetError(f"函数 `{stmt.name}` 重复定义")
                self.functions[stmt.name] = stmt.ret

    # ---- 顶层 ----

    def emit_program(self, program: ast.Program) -> None:
        self.push("package main")
        self.push("")

        if program_uses_print(program):
            self.push('import "fmt"')
            self.push("")

        # 函数声明 → 包级 func
        for stmt in program.stmts:
            if isinstance(stmt, ast.FnDecl):
                self.emit_stmt(stmt)
                self.push("")

        # 其余顶层语句 → main
        self.push("func main() {")
        self.indent += 1
        for stmt in program.stmts:
            if not isinstance(stmt, ast.FnDecl):
                self.emit_stmt(stmt)
        self.indent -= 1
        self.push("}")

    # ---- 语句 ----

    def emit_stmt(self, stmt: ast.Stmt) -> None:
        if isinstance(stmt, ast.Let):
            t = stmt.ty
            if t is None:
                t = self.infer_type(stmt.value)
                if t is None:
                    raise TenetError(f"无法推断 `{stmt.name}` 的类型，请显式标注")
            self.symbols[stmt.name] = t
            v = self.emit_expr(stmt.value)
            self.push(f"var {stmt.name} {go_type(t)} = {v};")
            return

        if isinstance(stmt, ast.ExprStmt):
            expr = stmt.expr
            if isinstance(expr, ast.Call):
                args = ", ".join(self.emit_expr(a) for a in expr.args)
                if expr.callee == "print":
                    self.push(f"fmt.Println({args});")
                else:
                    self.push(f"{expr.callee}({args});")
                return
            if isinstance(expr, ast.Assign):
                v = self.emit_expr(expr.value)
                self.push(f"{expr.name} = {v};")
                return
            raise TenetError("语句必须是函数调用或赋值")

        if isinstance(stmt, ast.If):
            self.emit_if(stmt.cond, stmt.then_branch, stmt.else_branch, "")
            return

        if isinstance(stmt, ast.While):
            c = self.emit_expr(stmt.cond)
            self.push(f"for {c} {{")
            self.indent += 1
            for s in stmt.body:
                self.emit_stmt(s)
            self.indent -= 1
            self.push("}")
            return

        if isinstance(stmt, ast.Return):
            if stmt.expr is not None:
                v = self.emit_expr(stmt.expr)
                self.push(f"return {v};")
            else:
                self.push("return;")
            return

        if isinstance(stmt, ast.Break):
            self.push("break;")
            return

        if isinstance(stmt, ast.Block):
            self.push("{")
            self.indent += 1
            for s in stmt.stmts:
                self.emit_stmt(s)
            self.indent -= 1
            self.push("}")
            return

        if isinstance(stmt, ast.FnDecl):
            sig_params = ", ".join(f"{n} {go_type(t)}" for n, t in stmt.params)
            ret = go_type(stmt.ret) if stmt.ret else ""
            if ret:
                sig = f"{stmt.name}({sig_params}) {ret}"
            else:
                sig = f"{stmt.name}({sig_params})"
            self.push(f"func {sig} {{")
            self.indent += 1
            # 参数进入符号表（函数体结束后恢复）
            saved = dict(self.symbols)
            for n, t in stmt.params:
                self.symbols[n] = t
            for s in stmt.body:
                self.emit_stmt(s)
            self.indent -= 1
            self.push("}")
            self.symbols = saved
            return

        raise TenetError(f"未知的语句类型: {type(stmt).__name__}")

    def emit_if(self, cond, then_branch, else_branch, prefix: str) -> None:
        c = self.emit_expr(cond)
        self.push(f"{prefix}if {c} {{")
        self.indent += 1
        for s in then_branch:
            self.emit_stmt(s)
        self.indent -= 1
        if else_branch is not None:
            # else if 链：单条 if 语句的 else 分支 → Go 的 `} else if ...`
            if len(else_branch) == 1 and isinstance(else_branch[0], ast.If):
                nested = else_branch[0]
                self.emit_if(nested.cond, nested.then_branch, nested.else_branch, "} else ")
            else:
                self.push("} else {")
                self.indent += 1
                for s in else_branch:
                    self.emit_stmt(s)
                self.indent -= 1
                self.push("}")
        else:
            self.push("}")

    # ---- 表达式 ----

    def emit_expr(self, expr: ast.Expr) -> str:
        if isinstance(expr, ast.IntLit):
            return str(expr.value)
        if isinstance(expr, ast.FloatLit):
            # Rust 的 f64 Display 会把 2.0 打印成 "2"，需要补回小数点，
            # 否则 Go 里 7 / 2.0 会变成整数除法 7 / 2
            s = val.fmt_float(expr.value)
            if "." not in s:
                s += ".0"
            return s
        if isinstance(expr, ast.StrLit):
            return go_string(expr.value)
        if isinstance(expr, ast.BoolLit):
            return "true" if expr.value else "false"
        if isinstance(expr, ast.Var):
            return expr.name
        if isinstance(expr, ast.Assign):
            return f"{expr.name} = {self.emit_expr(expr.value)}"
        if isinstance(expr, ast.Unary):
            inner = self.emit_expr(expr.expr)
            return f"-{inner}" if expr.op == ast.OP_NEG else f"!{inner}"
        if isinstance(expr, ast.Binary):
            l = self.emit_expr(expr.lhs)
            r = self.emit_expr(expr.rhs)
            return f"({l} {go_binop(expr.op)} {r})"
        if isinstance(expr, ast.Call):
            args = ", ".join(self.emit_expr(a) for a in expr.args)
            if expr.callee == "print":
                return f"fmt.Println({args})"
            return f"{expr.callee}({args})"
        raise TenetError(f"未知的表达式类型: {type(expr).__name__}")

    # ---- 类型推断（与解释器语义对齐）----

    def infer_type(self, expr: ast.Expr) -> str | None:
        if isinstance(expr, ast.IntLit):
            return ast.T_INT
        if isinstance(expr, ast.FloatLit):
            return ast.T_FLOAT
        if isinstance(expr, ast.StrLit):
            return ast.T_STR
        if isinstance(expr, ast.BoolLit):
            return ast.T_BOOL
        if isinstance(expr, ast.Var):
            return self.symbols.get(expr.name)
        if isinstance(expr, ast.Assign):
            return self.symbols.get(expr.name)
        if isinstance(expr, ast.Unary):
            if expr.op == ast.OP_NOT:
                return ast.T_BOOL
            t = self.infer_type(expr.expr)
            if t in (ast.T_INT, ast.T_FLOAT):
                return t
            return None
        if isinstance(expr, ast.Binary):
            lt = self.infer_type(expr.lhs)
            rt = self.infer_type(expr.rhs)
            op = expr.op
            if op in (ast.OP_AND, ast.OP_OR):
                return ast.T_BOOL
            if op in (ast.OP_EQ, ast.OP_NEQ, ast.OP_LT, ast.OP_LTE, ast.OP_GT, ast.OP_GTE):
                return ast.T_BOOL
            if op == ast.OP_MOD:
                return ast.T_INT if lt == ast.T_INT and rt == ast.T_INT else None
            if op == ast.OP_ADD:
                if lt == ast.T_STR and rt == ast.T_STR:
                    return ast.T_STR
                if lt == ast.T_FLOAT or rt == ast.T_FLOAT:
                    return ast.T_FLOAT
                if lt == ast.T_INT and rt == ast.T_INT:
                    return ast.T_INT
                return None
            # Sub / Mul / Div
            if lt == ast.T_FLOAT or rt == ast.T_FLOAT:
                return ast.T_FLOAT
            if lt == ast.T_INT and rt == ast.T_INT:
                return ast.T_INT
            return None
        if isinstance(expr, ast.Call):
            return self.functions.get(expr.callee)
        return None


def go_string(s: str) -> str:
    """把 Tenet 字符串转成 Go 字符串字面量。"""
    out = ['"']
    for c in s:
        if c == '"':
            out.append('\\"')
        elif c == "\\":
            out.append("\\\\")
        elif c == "\n":
            out.append("\\n")
        elif c == "\t":
            out.append("\\t")
        elif c == "\r":
            out.append("\\r")
        elif ord(c) < 0x20:
            out.append(f"\\x{ord(c):02x}")
        else:
            out.append(c)
    out.append('"')
    return "".join(out)


def program_uses_print(program: ast.Program) -> bool:
    """扫描整个程序是否调用 `print`（决定是否生成 `import "fmt"`）。"""

    def stmt_uses_print(stmts) -> bool:
        return any(stmt_uses_print_one(s) for s in stmts)

    def stmt_uses_print_one(stmt) -> bool:
        if isinstance(stmt, (ast.ExprStmt, ast.Let, ast.Return)):
            if isinstance(stmt, ast.ExprStmt):
                return expr_uses_print(stmt.expr)
            if isinstance(stmt, ast.Let):
                return expr_uses_print(stmt.value)
            return stmt.expr is not None and expr_uses_print(stmt.expr)
        if isinstance(stmt, ast.If):
            return (
                expr_uses_print(stmt.cond)
                or stmt_uses_print(stmt.then_branch)
                or (stmt.else_branch is not None and stmt_uses_print(stmt.else_branch))
            )
        if isinstance(stmt, ast.While):
            return expr_uses_print(stmt.cond) or stmt_uses_print(stmt.body)
        if isinstance(stmt, ast.FnDecl):
            return stmt_uses_print(stmt.body)
        if isinstance(stmt, ast.Block):
            return stmt_uses_print(stmt.stmts)
        return False

    def expr_uses_print(expr) -> bool:
        if isinstance(expr, ast.Call):
            return expr.callee == "print" or any(expr_uses_print(a) for a in expr.args)
        if isinstance(expr, ast.Assign):
            return expr_uses_print(expr.value)
        if isinstance(expr, ast.Unary):
            return expr_uses_print(expr.expr)
        if isinstance(expr, ast.Binary):
            return expr_uses_print(expr.lhs) or expr_uses_print(expr.rhs)
        return False

    return stmt_uses_print(program.stmts)

File: tenet/impl-rs/src/codegen.rs
//! 代码生成器（Codegen）：把 Tenet AST 编译成 Go 源码。
//!
//! 这是「源码到源码」的编译器前端：Tenet 是静态类型语言，因此
//! 可以一一映射到 Go——类型、语句、表达式几乎逐行对应：
//!
//! | Tenet | Go |
//! |-------|----|
//! | `int` / `float` / `bool` / `string` | `int64` / `float64` / `bool` / `string` |
//! | `let x: T = v;` | `var x T = v;` |
//! | `while (c) { }` | `for c { }`（Go 没有 while） |
//! | `print(...)` | `fmt.Println(...)` |
//! | `fn f(a: int) -> int { }` | `func f(a int64) int64 { }` |
//!
//! `let` 省略类型标注时，由 `infer_type` 静态推断（字面量、变量、
//! 运算规则、函数返回类型），与解释器的动态语义保持一致。

use std::collections::HashMap;

use crate::ast::{BinaryOp, Expr, Program, Stmt, Type, UnaryOp};
use crate::error::{TenetError, TResult};

/// 函数签名（代码生成阶段只需要返回类型用于推断）。
struct FnSig {
    ret: Option<Type>,
}

pub struct GoCodegen {
    out: String,
    indent: usize,
    /// 当前作用域内可见的变量类型。
    symbols: HashMap<String, Type>,
    /// 全局函数签名表。
    functions: HashMap<String, FnSig>,
}

impl GoCodegen {
    pub fn new() -> Self {
        Self {
            out: String::new(),
            indent: 0,
            symbols: HashMap::new(),
            functions: HashMap::new(),
        }
    }

    /// 顶层入口：整个程序 → Go 源码字符串。
    pub fn generate(program: &Program) -> TResult<String> {
        let mut gen = GoCodegen::new();
        gen.collect_signatures(program)?;
        gen.emit_program(program)?;
        Ok(gen.out)
    }

    // ---- 预处理 ----

    /// 第一遍：收集所有函数签名（用于调用表达式推断返回类型）。
    fn collect_signatures(&mut self, program: &Program) -> TResult<()> {
        for stmt in &program.stmts {
            if let Stmt::FnDecl {
                name,
                ret,
                ..
            } = stmt
            {
                if self.functions.contains_key(name) {
                    return Err(TenetError::new(format!("函数 `{name}` 重复定义"), None));
                }
                self.functions.insert(
                    name.clone(),
                    FnSig {
                        ret: *ret,
                    },
                );
            }
        }
        Ok(())
    }

    // ---- 顶层 ----

    fn emit_program(&mut self, program: &Program) -> TResult<()> {
        self.push("package main");
        self.push("");

        // 是否用到 print（决定是否 import fmt）
        if program_uses_print(program) {
            self.push("import \"fmt\"");
            self.push("");
        }

        // 函数声明 → 包级 func
        for stmt in &program.stmts {
            if let Stmt::FnDecl { .. } = stmt {
                self.emit_stmt(stmt)?;
                self.push("");
            }
        }

        // 其余顶层语句 → main
        self.push("func main() {");
        self.indent += 1;
        for stmt in &program.stmts {
            if !matches!(stmt, Stmt::FnDecl { .. }) {
                self.emit_stmt(stmt)?;
            }
        }
        self.indent -= 1;
        self.push("}");
        Ok(())
    }

    // ---- 语句 ----

    fn emit_stmt(&mut self, stmt: &Stmt) -> TResult<()> {
        match stmt {
            Stmt::Let { name, ty, value } => {
                let t = match ty {
                    Some(t) => *t,
                    None => self
                        .infer_type(value)?
                        .ok_or_else(|| {
                            TenetError::new(format!("无法推断 `{name}` 的类型，请显式标注"), None)
                        })?,
                };
                self.symbols.insert(name.clone(), t);
                let val = self.emit_expr(value)?;
                self.push(&format!("var {name} {} = {val};", go_type(t)));
            }
            Stmt::Expr(expr) => match expr {
                Expr::Call { callee, args } => {
                    let args = self.emit_args(args)?;
                    if callee == "print" {
                        self.push(&format!("fmt.Println({});", args.join(", ")));
                    } else {
                        self.push(&format!("{callee}({});", args.join(", ")));
                    }
                }
                Expr::Assign { name, value } => {
                    let val = self.emit_expr(value)?;
                    self.push(&format!("{name} = {val};"));
                }
                _ => return Err(TenetError::new("语句必须是函数调用或赋值", None)),
            },
            Stmt::If {
                cond,
                then_branch,
                else_branch,
            } => self.emit_if(cond, then_branch, else_branch, "")?,
            Stmt::While { cond, body } => {
                let c = self.emit_expr(cond)?;
                self.push(&format!("for {c} {{"));
                self.indent += 1;
                for s in body {
                    self.emit_stmt(s)?;
                }
                self.indent -= 1;
                self.push("}");
            }
            Stmt::Return(expr) => match expr {
                Some(e) => {
                    let val = self.emit_expr(e)?;
                    self.push(&format!("return {val};"));
                }
                None => self.push("return;"),
            },
            Stmt::Break => self.push("break;"),
            Stmt::Block(stmts) => {
                self.push("{");
                self.indent += 1;
                for s in stmts {
                    self.emit_stmt(s)?;
                }
                self.indent -= 1;
                self.push("}");
            }
            Stmt::FnDecl {
                name,
                params,
                ret,
                body,
            } => {
                let sig_params: Vec<String> =
                    params.iter().map(|(n, t)| format!("{n} {}", go_type(*t))).collect();
                let ret = ret.map(go_type).unwrap_or_default();
                let sig = if ret.is_empty() {
                    format!("{name}({})", sig_params.join(", "))
                } else {
                    format!("{name}({}) {ret}", sig_params.join(", "))
                };
                self.push(&format!("func {sig} {{"));
                self.indent += 1;
                // 参数进入符号表（函数体结束后恢复）
                let saved = self.symbols.clone();
                for (n, t) in params {
                    self.symbols.insert(n.clone(), *t);
                }
                for s in body {
                    self.emit_stmt(s)?;
                }
                self.indent -= 1;
                self.push("}");
                self.symbols = saved;
            }
        }
        Ok(())
    }

    /// 发射 if 语句；`prefix` 用于 `} else if` 的续行拼接。
    fn emit_if(
        &mut self,
        cond: &Expr,
        then_branch: &[Stmt],
        else_branch: &Option<Vec<Stmt>>,
        prefix: &str,
    ) -> TResult<()> {
        let c = self.emit_expr(cond)?;
        self.push(&format!("{prefix}if {c} {{"));
        self.indent += 1;
        for s in then_branch {
            self.emit_stmt(s)?;
        }
        self.indent -= 1;
        match else_branch {
            // else if 链：单条 if 语句的 else 分支 → Go 的 `} else if ...`
            Some(branch) if branch.len() == 1 && matches!(branch[0], Stmt::If { .. }) => {
                if let Stmt::If {
                    cond,
                    then_branch,
                    else_branch,
                } = &branch[0]
                {
                    self.emit_if(cond, then_branch, else_branch, "} else ")?;
                }
            }
            Some(branch) => {
                self.push("} else {");
                self.indent += 1;
                for s in branch {
                    self.emit_stmt(s)?;
                }
                self.indent -= 1;
                self.push("}");
            }
            None => self.push("}"),
        }
        Ok(())
    }

    // ---- 表达式 ----

    fn emit_expr(&mut self, expr: &Expr) -> TResult<String> {
        Ok(match expr {
            Expr::Int(v) => v.to_string(),
            Expr::Float(v) => {
                // Rust 的 f64 Display 会把 2.0 打印成 "2"，会丢失浮点语义，
                // 因此强制补上小数点：2.0 → "2.0"
                let s = v.to_string();
                if s.contains('.') {
                    s
                } else {
                    format!("{s}.0")
                }
            }
            Expr::Str(s) => go_string(s),
            Expr::Bool(b) => b.to_string(),
            Expr::Var(name) => name.clone(),
            Expr::Assign { name, value } => {
                let val = self.emit_expr(value)?;
                format!("{name} = {val}")
            }
            Expr::Unary { op, expr } => {
                let inner = self.emit_expr(expr)?;
                match op {
                    UnaryOp::Neg => format!("-{inner}"),
                    UnaryOp::Not => format!("!{inner}"),
                }
            }
            Expr::Binary { op, lhs, rhs } => {
                let l = self.emit_expr(lhs)?;
                let r = self.emit_expr(rhs)?;
                format!("({l} {} {r})", go_binop(*op))
            }
            Expr::Call { callee, args } => {
                let args = self.emit_args(args)?;
                if callee == "print" {
                    format!("fmt.Println({})", args.join(", "))
                } else {
                    format!("{callee}({})", args.join(", "))
                }
            }
        })
    }

    fn emit_args(&mut self, args: &[Expr]) -> TResult<Vec<String>> {
        args.iter().map(|a| self.emit_expr(a)).collect()
    }

    // ---- 类型推断（与解释器语义对齐）----

    fn infer_type(&self, expr: &Expr) -> TResult<Option<Type>> {
        Ok(match expr {
            Expr::Int(_) => Some(Type::Int),
            Expr::Float(_) => Some(Type::Float),
            Expr::Str(_) => Some(Type::Str),
            Expr::Bool(_) => Some(Type::Bool),
            Expr::Var(name) => self.symbols.get(name).copied(),
            Expr::Assign { name, .. } => self.symbols.get(name).copied(),
            Expr::Unary { op, expr } => match op {
                UnaryOp::Not => Some(Type::Bool),
                UnaryOp::Neg => match self.infer_type(expr)? {
                    Some(Type::Int) => Some(Type::Int),
                    Some(Type::Float) => Some(Type::Float),
                    _ => None,
                },
            },
            Expr::Binary { op, lhs, rhs } => {
                let lt = self.infer_type(lhs)?;
                let rt = self.infer_type(rhs)?;
                match op {
                    BinaryOp::And | BinaryOp::Or => Some(Type::Bool),
                    BinaryOp::Eq
                    | BinaryOp::NotEq
                    | BinaryOp::Lt
                    | BinaryOp::LtEq
                    | BinaryOp::Gt
                    | BinaryOp::GtEq => Some(Type::Bool),
                    BinaryOp::Mod => {
                        if lt == Some(Type::Int) && rt == Some(Type::Int) {
                            Some(Type::Int)
                        } else {
                            None
                        }
                    }
                    BinaryOp::Add => match (lt, rt) {
                        (Some(Type::Str), Some(Type::Str)) => Some(Type::Str),
                        (Some(Type::Float), _) | (_, Some(Type::Float)) => Some(Type::Float),
                        (Some(Type::Int), Some(Type::Int)) => Some(Type::Int),
                        _ => None,
                    },
                    BinaryOp::Sub | BinaryOp::Mul | BinaryOp::Div => match (lt, rt) {
                        (Some(Type::Float), _) | (_, Some(Type::Float)) => Some(Type::Float),
                        (Some(Type::Int), Some(Type::Int)) => Some(Type::Int),
                        _ => None,
                    },
                }
            }
            Expr::Call { callee, .. } => self.functions.get(callee).and_then(|s| s.ret),
        })
    }
}

// ---- 输出辅助 ----

impl GoCodegen {
    fn push(&mut self, line: &str) {
        if line.is_empty() {
            self.out.push('\n');
            return;
        }
        for _ in 0..self.indent {
            self.out.push('\t');
        }
        self.out.push_str(line);
        self.out.push('\n');
    }
}

// ---- 映射辅助 ----

fn go_type(t: Type) -> &'static str {
    match t {
        Type::Int => "int64",
        Type::Float => "float64",
        Type::Bool => "bool",
        Type::Str => "string",
    }
}

fn go_binop(op: BinaryOp) -> &'static str {
    match op {
        BinaryOp::Add => "+",
        BinaryOp::Sub => "-",
        BinaryOp::Mul => "*",
        BinaryOp::Div => "/",
        BinaryOp::Mod => "%",
        BinaryOp::Eq => "==",
        BinaryOp::NotEq => "!=",
        BinaryOp::Lt => "<",
        BinaryOp::LtEq => "<=",
        BinaryOp::Gt => ">",
        BinaryOp::GtEq => ">=",
        BinaryOp::And => "&&",
        BinaryOp::Or => "||",
    }
}

/// 把 Tenet 字符串转成 Go 字符串字面量。
fn go_string(s: &str) -> String {
    let mut out = String::from("\"");
    for c in s.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\t' => out.push_str("\\t"),
            '\r' => out.push_str("\\r"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\x{:02x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
    out
}

/// 扫描整个程序是否调用 `print`（决定是否生成 `import "fmt"`）。
fn program_uses_print(program: &Program) -> bool {
    fn stmt_uses_print(stmts: &[Stmt]) -> bool {
        stmts.iter().any(stmt_uses_print_one)
    }
    fn stmt_uses_print_one(stmt: &Stmt) -> bool {
        match stmt {
            Stmt::Expr(expr) | Stmt::Let { value: expr, .. } | Stmt::Return(Some(expr)) => {
                expr_uses_print(expr)
            }
            Stmt::If {
                cond,
                then_branch,
                else_branch,
            } => {
                expr_uses_print(cond)
                    || stmt_uses_print(then_branch)
                    || else_branch.as_deref().map_or(false, stmt_uses_print)
            }
            Stmt::While { cond, body } => expr_uses_print(cond) || stmt_uses_print(body),
            Stmt::FnDecl { body, .. } => stmt_uses_print(body),
            Stmt::Block(stmts) => stmt_uses_print(stmts),
            Stmt::Return(None) | Stmt::Break => false,
        }
    }
    fn expr_uses_print(expr: &Expr) -> bool {
        match expr {
            Expr::Call { callee, args } => callee == "print" || args.iter().any(expr_uses_print),
            Expr::Assign { value, .. } => expr_uses_print(value),
            Expr::Unary { expr, .. } => expr_uses_print(expr),
            Expr::Binary { lhs, rhs, .. } => expr_uses_print(lhs) || expr_uses_print(rhs),
            _ => false,
        }
    }
    stmt_uses_print(&program.stmts)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::Parser;

    fn gen(src: &str) -> TResult<String> {
        let prog = Parser::parse(src)?;
        GoCodegen::generate(&prog)
    }

    #[test]
    fn hello_world() {
        let out = gen("print(\"hello\");").unwrap();
        assert!(out.contains("package main"), "{out}");
        assert!(out.contains("import \"fmt\""), "{out}");
        assert!(out.contains("func main()"), "{out}");
        assert!(out.contains("fmt.Println(\"hello\")"), "{out}");
    }

    #[test]
    fn no_print_no_fmt_import() {
        let out = gen("let x: int = 1;").unwrap();
        assert!(!out.contains("import"), "{out}");
    }

    #[test]
    fn let_with_annotation() {
        let out = gen("let x: int = 42; let s: string = \"hi\";").unwrap();
        assert!(out.contains("var x int64 = 42;"), "{out}");
        assert!(out.contains("var s string = \"hi\";"), "{out}");
    }

    #[test]
    fn let_type_inference() {
        let out = gen("let x = 1 + 2.5;").unwrap();
        assert!(out.contains("var x float64 = (1 + 2.5);"), "{out}");
    }

    #[test]
    fn inference_from_function_return() {
        let out = gen("fn f() -> int { return 1; } let x = f();").unwrap();
        assert!(out.contains("var x int64 = f();"), "{out}");
    }

    #[test]
    fn while_becomes_for() {
        let out = gen("while (x < 10) { x = x + 1; }").unwrap();
        assert!(out.contains("for (x < 10) {"), "{out}");
    }

    #[test]
    fn function_decl() {
        let out = gen("fn add(a: int, b: int) -> int { return a + b; }").unwrap();
        assert!(out.contains("func add(a int64, b int64) int64 {"), "{out}");
        assert!(out.contains("return (a + b);"), "{out}");
    }

    #[test]
    fn void_function() {
        let out = gen("fn greet() { print(\"hi\"); }").unwrap();
        assert!(out.contains("func greet() {"), "{out}");
        assert!(!out.contains("func greet()  {"), "{out}");
    }

    #[test]
    fn else_if_chain() {
        let out = gen(
            "if (x > 0) { print(1); } else if (x < 0) { print(-1); } else { print(0); }",
        )
        .unwrap();
        assert!(out.contains("} else if (x < 0) {"), "{out}");
        assert!(out.contains("} else {"), "{out}");
    }

    #[test]
    fn string_escaping() {
        let out = gen("print(\"a\\n\\\"b\\\"\");").unwrap();
        assert!(out.contains("fmt.Println(\"a\\n\\\"b\\\"\")"), "{out}");
    }

    #[test]
    fn print_inside_function_triggers_import() {
        let out = gen("fn f() { print(1); }").unwrap();
        assert!(out.contains("import \"fmt\""), "{out}");
    }

    #[test]
    fn float_literal_keeps_decimal_point() {
        // Rust 的 f64 Display 会把 2.0 打印成 "2"，代码生成必须补回小数点，
        // 否则 Go 里 7 / 2.0 会变成整数除法 7 / 2
        let out = gen("print(7 / 2.0);").unwrap();
        assert!(out.contains("(7 / 2.0)"), "{out}");
    }

    #[test]
    fn cannot_infer_type_errors() {
        // 无返回类型的函数调用无法推断类型
        let err = gen("fn f() { } let x = f();").unwrap_err();
        assert!(err.message.contains("无法推断"), "{}", err);
    }
}
